plugins/devlink/README.md

```diff
@@ -0,0 +1,110 @@
+# Devlink Plugin
+
+This plugin provides firmware update support for network interface cards that support the Linux devlink interface. This is a generic plugin that can work with any device that implements devlink functionality.
+
+## Supported Devices
+
+The plugin supports any device that implements the devlink interface, regardless the bus it resides on.
+
+## How It Works
+
+The plugin uses the Linux devlink netlink interface to communicate with the kernel and perform firmware updates. The process involves:
+
+1. **Netlink Communication**: Opens Netlink socket to communicate with the devlink subsystem
+2. **Device Detection**:  Gets all existing devlink devices
+3. **Firmware Upload**: Writes the firmware file to `/lib/firmware/` and instructs devlink to flash it
+4. **Progress Monitoring**: Monitors devlink status messages to provide real-time progress updates
+5. **Firmware Activation**: Activates firmware using devlink reload activate action
+
+## Requirements
+
+- Linux kernel with devlink support
+- Device driver with devlink implementation
+- Root privileges (required for devlink write commands)
+
+## Usage
+
+The plugin integrates seamlessly with fwupd. Once installed, supported devices will appear in the device list:
+
+```bash
+# List devices
+fwupdmgr get-devices
+
+# Update firmware
+fwupdmgr update
+```
+
+## Technical Details
+
+### Devlink Protocol
+
+The plugin implements the devlink generic netlink protocol:
+
+1. **Device Enumeration**: Sends `DEVLINK_CMD_GET` with dump flag to discover all devlink devices
+2. **Device Monitoring**: Receives `DEVLINK_CMD_NEW` notifications when devices are added
+3. **Device Removal**: Receives `DEVLINK_CMD_DEL` notifications when devices are removed
+4. **Device Information**: Sends `DEVLINK_CMD_INFO_GET` to retrieve device details and versions
+5. **Flash Command**: Sends `DEVLINK_CMD_FLASH_UPDATE` with device and file information
+6. **Status Monitoring**: Receives `DEVLINK_CMD_FLASH_UPDATE_STATUS` messages for progress
+7. **Completion**: Waits for `DEVLINK_CMD_FLASH_UPDATE_END` to confirm completion
+8. **Firmware Activation**: Sends `DEVLINK_CMD_RELOAD` with fw_activate action to activate updated firmware
+
+### Private Flags
+
+The plugin supports the following private flags:
+
+#### `omit-component-name`
+
+When this flag is set, the plugin will not include the `DEVLINK_ATTR_FLASH_UPDATE_COMPONENT` attribute in the flash command. This is useful for devices that don't require or support component-specific updates.
+
+**Usage in metainfo XML:**
+
+```xml
+<custom>
+  <value key="LVFS::DeviceFlags">omit-component-name</value>
```

Comment:
do we want this in the metadata, or just in the quirk file? i.e. is there going to be a quirk entry for each device supported by the devlink plugin (opt in) or is it going to be fairly promiscuous in reality?

Reply:
My idea is to have the devlink plugin as generic as possible and possibly without any quirks. Then, it is up the the firmware provider to specify what are his device needs. If I issue FW cab for mlx5, I know I need to set this flag. Intel ice firmware provider knows he does not need this. Very simple, no need to do anything in fwupd for other devlink driver users.

Reply:
Sure, that's probably sane. Is there any hardware I should obtain so I test this myself? e.g. buy something like a ConnectX-5 to debug with?

Reply:
> is currently not available on public web

Could you sent me some loaner hardware and firmware using the existing Red Hat <=> NVIDIA NDA in place?